src/NUnitFramework/framework/Attributes/TestFixtureSourceAttribute.cs

```diff
@@ -111,7 +111,7 @@ public IEnumerable<TestSuite> BuildFrom(ITypeInfo typeInfo)
         {
             Type sourceType = SourceType ?? typeInfo.Type;
 
-            foreach (TestFixtureParameters parms in GetParametersFor(sourceType))
+            foreach (ITestFixtureData parms in GetParametersFor(sourceType))
```

Comment:
This change needs more discussion - see https://github.com/nunit/nunit/issues/2019, where we decided that users were required to use the TestFixtureData class, rather than the interface. Re-reading that issue, I think I would still be in favour of keeping TestFixtureData inline with TestCaseData - if the latter is harder to change. @StanEgo - did you have a particular reason for wanting this functionality, or did you just assume it should work? I imagine the code/comments could be mis-leading here.

Reply:
Good catch on this. It looks like the docs still haven't been updated for either [TestCaseData](https://github.com/nunit/docs/wiki/TestCaseData) or [TestFixtureData](https://github.com/nunit/docs/wiki/TestFixtureData) which state 

> Although any object implementing IFixtureData may be used to provide specific instance information for a parameterized fixture, NUnit provides the TestFixtureData class for this purpose.

The docs should be updated appropriately. I agree that this should match `TestCaseData`. I will update the documentation now.

Reply:
The documentation has been updated. Please revert this change and then I think we are good to go.

Reply:
@rprouse Strictly speaking, we decided to require `TestFixtureParameters` and `TestCaseParameters`. The corresponding `TestXxxxData` classes are the same with the addition of a fluent interface.

Reply:
I see this as a bug in any case. If you want to allow TestFixtureParameters only, then this should  be controlled in builder, throwing proper exception and having relevant tests. Right now pretty strange class is responsible with very strange checking algorithm. Between GetParametersFor method that returns IEnumerable&lt;ITestFixtureData&gt; and NUnitTestFixtureBuilder.BuildFrom method that accepts ITestFixtureData as well, we insert redundant double casting that throws absolutely unclear InvalidCaseException for the user. Also no tests if this is an expectable behavior.

On the other side I believe that current ITestFixtureData-based design is absolutely SOLID. We have an SRP-interface to access test arguments and implementations those can fetch these arguments from any type of input. I don’t think that having only one implementation like TestFixtureParameters that can accepts object[] or another ITestFixtureData is a scalable solution. (btw, let's then also remove the constructor with ITestFixtureData, because it doesn't make sense to create one TestFixtureParameters from another).

In my initial message there is a sample, how this scalability was able to solve problem in existing code. This point of extension can be also used for helpers (when some complex object converted into array of args), args reusing (e.g. combining existing args and other set-of-args operations), conditions (depends on runtime environment), dispatching, proxying, multiplexing, there are millions of different cases. E.g. right now I have in one project solution like:

```
[Feature(typeof(IDeleteCommand<>))]
public class DeleteNonExistingSpecification<TEntity> {
	public override void Then(Exception exception) => Assert.That(
		exception,
		Is.InstanceOf<EntityNotFoundException>()
	);
}
```

That applies one specification to all IDeleteCommand<> implementations that registered in IoC. And FeatureAttribute works like a FixtureSource fetching types from IoC, but I had to write it from scratch, because currenly it is not extensible.

But I don't think that this is a question of trying to forecast all possible usages. The question is, why having a good design we don’t use it?

Explanation in #2019 sounds for me like: "If there is no easy fix visible, let's suppose that there is no problem" :smiley:

But probably this fix should be extrapolated to TestCaseParameters too, that looks not so easy, but possible.

Reply:
@StanEgo I think what's missing in the discussion among the developers and therefore invisible to you, is that we previously decided to make `TestFixtureParameters` part of the interface. As you point out, the decision was not implemented properly.

There's nothing wrong with an interface that specifies a particular data class that must be returned and that was the intention. Obviously, the existing `GetParametersFor` doesn't carry out the decision correctly.

I think the team has to make the whole thing consistent in some way... either going back to accepting the interface without error or moving ahead to using the data class in the interface.